File: Anton-Kraev/Term4/ProducerConsumer/ProducerConsumer/Producer.cs
namespace ProducerConsumer;

public class Producer<T>
{
    private readonly Mutex mutex;
    private readonly Buffer<T> buffer;
    private readonly Func<T> produce;
    private volatile bool isStopped;

    public Producer(Mutex mutex, Buffer<T> buffer, Func<T> produce)
    {
        this.mutex = mutex;
        this.buffer = buffer;
        this.produce = produce;
    }

    public void Run()
    {
        while (!isStopped)
        {
            var item = produce();

            mutex.WaitOne();
            buffer.Push(item);
            mutex.ReleaseMutex();

            Thread.Sleep(10);
        }
    }

    public void Stop()
    {
        isStopped = true;
    }
}

File: Khelmianov-Artem/Term4/Block2/ProducerConsumer/src/main/kotlin/Producer.kt
import java.util.concurrent.locks.ReentrantLock
import kotlin.random.Random

class Producer(
    queue: MutableList<Data>,
    mutex: ReentrantLock,
    name: String = ""
) : Actor<Data>(queue, mutex, name) {
    private var cnt = 0

    override fun run() {
        isRunning = true
        while (isRunning) {
            val item = Data("$name-${cnt++}")
            mutex.withLock {
                if (!isRunning) return@withLock
                queue.add(item)
                println("$name produced $item")
            }
            Thread.sleep(Random.nextLong(1000, 2000))
        }
    }
}
